__factories__/clients/firebase.js

```diff
@@ -0,0 +1,31 @@
+import {Factory} from 'rosie';
+
+class FirebaseError extends Error {
+  constructor(args) {
+    super(args.name);
+    this.code = args.name;
+    this.credential = args.credential;
+    Error.captureStackTrace(this, FirebaseError);
+  }
+}
+
+export const credential = new Factory().attrs({
+  providerId: 'github.com',
+  accessToken: 'abc123',
+});
+
+export const githubProfile = new Factory().attrs({
+  login: 'popcoder',
+});
+
+export const firebaseError = Factory.define(
+  'firebaseError',
+  FirebaseError,
+).attrs({
+  name: 'some other error',
+});
+
+export const credentialInUseError = new Factory().extend(firebaseError).attrs({
+  name: 'auth/credential-already-in-use',
+  credential: credential.build(),
```

Comment:
I think when an attribute is a full object (rather than a scalar value) we want to build it at instantiation time; otherwise all `credentialInUseError`s will hold a reference to the _same_ credential object, meaning mutating it in one place would mutate it everywhere. Of course running into that sort of problem is not practically very likely, but still better to establish a practice of avoiding it altogether imo. So here I think you would just want to do:

```
attr('credential', () => credential.build())
```